Zotero/Scenes/AppCoordinator.swift

```diff
@@ -61,424 +59,331 @@ final class AppCoordinator: NSObject {
-
-        var url: URL?
-        var library: Library?
-        var collection: Collection?
-
-        do {
-            try dbStorage.perform(on: .main, with: { coordinator in
-                let item = try coordinator.perform(request: ReadItemDbRequest(libraryId: libraryId, key: key))
-
-                guard let attachment = AttachmentCreator.attachment(for: item, fileStorage: self.controllers.fileStorage, urlDetector: nil) else { return }
-
-                switch attachment.type {
-                case .file(let filename, let contentType, let location, _):
-                    switch location {
-                    case .local, .localAndChangedRemotely:
-                        let file = Files.attachmentFile(in: libraryId, key: key, filename: filename, contentType: contentType)
-                        url = file.createUrl()
-                        let (_collection, _library) = try coordinator.perform(request: ReadCollectionAndLibraryDbRequest(collectionId: collectionId, libraryId: libraryId))
-                        collection = _collection
-                        library = _library
+        switch customUrl {
+        case .itemDetail(let key, let library, let preselectedChildKey):
+            DDLogInfo("AppCoordinator: show custom url - item detail; key=\(key); library=\(library.identifier)")
+            showItemDetail(in: mainController, key: key, library: library, selectChildKey: preselectedChildKey, animated: animated, dismissIfPresenting: true)
 
-                    case .remote, .remoteMissing: break
+        case .pdfReader(let attachment, let library, let page, let annotation, let parentKey, let isAvailable):
+            let message = DDLogMessageFormat(
+                stringLiteral:
+                    "AppCoordinator: show custom url - pdf reader; key=\(attachment.key); library=\(library.identifier);" +
+                " page=\(page.flatMap(String.init) ?? "nil"); annotation=\(annotation ?? "nil"); parentKey=\(parentKey ?? "nil")"
+            )
+            DDLogInfo(message)
+            mainController.getDetailCoordinator { coordinator in
+                if isAvailable {
+                    guard (coordinator.navigationController?.presentedViewController as? PDFReaderViewController)?.key != attachment.key else { return }
+                    open(attachment: attachment, library: library, on: page, annotation: annotation, window: window, detailCoordinator: coordinator, animated: animated) {
+                        showItemDetail(in: mainController, key: (parentKey ?? attachment.key), library: library, selectChildKey: attachment.key, animated: animated, dismissIfPresenting: false)
+                    }
+                } else {
+                    showItemDetail(in: mainController, key: (parentKey ?? attachment.key), library: library, selectChildKey: attachment.key, animated: animated, dismissIfPresenting: true)
+                    download(attachment: attachment, parentKey: parentKey) {
```

Comment:
This probably won't have any effect, but just to be sure: when `showItemDetail` dismisses a controller from `mainController` we should probably add a completion handler and do `open(attachment:...)` only after item detail completed dismissing presented controller. Even though `download(attachment: parentKey:)` should provide sufficient delay. 

I guess if the device has no internet connection and we're presenting an attachment that's available locally but updated remotely, the download request will immediately abort and the app will try to present existing PDF. So there might actually be conflict in this case.

Reply:
Check https://github.com/zotero/zotero-ios/pull/826/commits/03a325a3638b7b9007c1ecc7eb9c4180310648e4